packages/bundle-source/cache.js

```diff
@@ -81,57 +86,91 @@ export const makeBundleCache = (wr, cwd, readPowers, opts) => {
-    const bundleWr = wr.neighbor(bundleFileName);
-    await bundleWr.writeText(code);
-    /** @type {import('fs').Stats} */
-    const { mtime: bundleTime } = await bundleWr.readOnly().stat();
 
-    /** @type {BundleMeta} */
-    const meta = {
-      bundleFileName,
-      bundleTime: bundleTime.toISOString(),
-      moduleSource: {
-        relative: bundleWr.readOnly().relative(srcRd.absolute()),
-        absolute: srcRd.absolute(),
-      },
-      contents: [...modTimeByPath.entries()].map(([relativePath, mtime]) => ({
-        relativePath,
-        mtime: mtime.toISOString(),
-      })),
-    };
+    const lockWr = wr.neighbor(toBundleLock(targetName));
 
-    await wr
-      .neighbor(toBundleMeta(targetName))
-      .writeText(JSON.stringify(meta, null, 2));
-    return meta;
+    // Check the bundle/meta file write lock.
+    try {
+      await lockWr.writeText('', { flag: 'wx' });
+    } catch (oops) {
+      if (oops.code !== 'EEXIST') {
+        throw oops;
+      }
+      // The lock exists, so something is already writing the bundle.
+      // eslint-disable-next-line no-use-before-define
+      return validate(targetName, rootPath);
+    }
+
+    try {
+      const bundleFileName = toBundleName(targetName);
+      const bundleWr = wr.neighbor(bundleFileName);
+      const metaWr = wr.neighbor(toBundleMeta(targetName));
+
+      // Prevent other processes from doing too much work just to see that we're
+      // already on it.
+      await metaWr.rm({ force: true });
+      await bundleWr.rm({ force: true });
+
+      const bundle = await bundleSource(rootPath, bundleOptions, {
+        ...readPowers,
+        read: loggedRead,
+      });
+
+      const { moduleFormat } = bundle;
+      assert.equal(moduleFormat, 'endoZipBase64');
+
+      const code = encodeBundle(bundle);
+      await wr.mkdir({ recursive: true });
+      await bundleWr.writeText(code);
+
+      /** @type {import('fs').Stats} */
+      const { mtime: bundleTime } = await bundleWr.readOnly().stat();
+
+      /** @type {BundleMeta} */
+      const meta = {
+        bundleFileName,
+        bundleTime: bundleTime.toISOString(),
+        moduleSource: {
+          relative: bundleWr.readOnly().relative(srcRd.absolute()),
+          absolute: srcRd.absolute(),
+        },
+        contents: [...modTimeByPath.entries()].map(([relativePath, mtime]) => ({
+          relativePath,
+          mtime: mtime.toISOString(),
+        })),
+      };
+
+      await metaWr.writeText(JSON.stringify(meta, null, 2));
+      return meta;
+    } finally {
+      await lockWr.rm({ force: true });
+    }
   };
 
-  const validate = async targetName => {
+  const validate = async (targetName, rootOpt, log = defaultLog) => {
     const metaRd = wr.readOnly().neighbor(toBundleMeta(targetName));
+    const lockRd = wr.readOnly().neighbor(toBundleLock(targetName));
+
+    // Wait for the bundle to be written.
+    // eslint-disable-next-line no-await-in-loop
```

Comment:
adding an `await null;` above the loop would make this await-safe, per recent discussion, yes?

Reply:
oh... the loop starts with an `await`. fair enough.
I wonder if a `for await ...` would be more clear.